lib/core/src/Cardano/Wallet/Primitive/Types.hs

```diff
@@ -876,32 +892,53 @@ slotDifference epl a b
     a' = flatSlot epl a
     b' = flatSlot epl b
 
+-- | Return the slot immediately before the given slot.
+slotPred :: SlotParameters -> SlotId -> SlotId
+slotPred (SlotParameters (EpochLength el) _ _) (SlotId en sn)
+    | sn > 0    = SlotId (en    ) (               sn - 1)
+    | otherwise = SlotId (en - 1) (fromIntegral $ el - 1)
+
+-- | Return the slot immediately after the given slot.
+slotSucc :: SlotParameters -> SlotId -> SlotId
+slotSucc (SlotParameters (EpochLength el) _ _) (SlotId en sn)
+    | sn < el - 1 = SlotId (en    ) (sn + 1)
+    | otherwise   = SlotId (en + 1) (     0)
+
 -- | The time that a slot begins.
-slotStartTime :: EpochLength -> SlotLength -> StartTime -> SlotId -> UTCTime
-slotStartTime epochLength (SlotLength slotLength) (StartTime start) sl =
-    addUTCTime offset start
+slotStartTime :: SlotParameters -> SlotId -> UTCTime
+slotStartTime (SlotParameters el (SlotLength sl) (StartTime st)) slot =
+    addUTCTime offset st
   where
-    offset = slotLength * fromIntegral (flatSlot epochLength sl)
-
--- | The SlotId at a given time.
-slotAt :: EpochLength -> SlotLength -> StartTime -> UTCTime -> SlotId
-slotAt (EpochLength nSlots) (SlotLength slotLength) (StartTime start) time =
-    SlotId
-        { epochNumber = ep
-        , slotNumber = sl
-        }
+    offset = sl * fromIntegral (flatSlot el slot)
+
+-- | For the given time 't', determine the ID of the earliest slot with start
+--   time 's' such that 't ≤ s'.
+slotStartingAtOrJustAfter :: SlotParameters -> UTCTime -> SlotId
+slotStartingAtOrJustAfter sp@(SlotParameters _ (SlotLength sl) _) t =
+    slotAt sp (addUTCTime (pred sl) t)
```

Comment:
we want to use `slotPred`, right?

Reply:
> we want to use `slotPred`, right?

I'm not sure that would work. What would you suggest?

These functions are connected with the following laws:
```hs
slotSucc . slotStartingAtOrJustBefore . utcTimePred . slotStartTime == id
slotPred . slotStartingAtOrJustAfter  . utcTimeSucc . slotStartTime == id

slotStartingAtOrJustBefore . slotStartTime == slotStartingAtOrJustAfter . slotStartTime
```